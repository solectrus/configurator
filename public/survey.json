{
  "locale": "de",
  "title": "SOLECTRUS-Konfigurator",
  "description": "Erstelle Schritt-für-Schritt deine Installation von SOLECTRUS",
  "logoPosition": "right",
  "pages": [
    {
      "name": "page1",
      "elements": [
        {
          "type": "checkbox",
          "name": "q_devices",
          "title": "Welche Geräte hast du in deinem Haus?",
          "defaultValue": ["battery"],
          "isRequired": true,
          "validators": [
            {
              "type": "expression",
              "text": "Ohne Stromspeicher geht's nicht!",
              "expression": "{q_devices} contains 'battery'"
            }
          ],
          "choices": [
            {
              "value": "battery",
              "text": "Stromspeicher"
            },
            {
              "value": "wallbox",
              "text": "Wallbox"
            },
            {
              "value": "heatpump",
              "text": "Wärmepumpe"
            }
          ]
        }
      ]
    },
    {
      "name": "page2",
      "elements": [
        {
          "type": "radiogroup",
          "name": "installation_type",
          "title": "Wie möchtest Du SOLECTRUS betreiben?",
          "isRequired": true,
          "choices": [
            {
              "value": "standalone",
              "text": "Alles auf einem lokalen Server (Raspberry Pi, NAS, sonstiger Linux-Server)"
            },
            {
              "value": "distributed",
              "text": "Verteilte Installation (Messwerte abfragen lokal, Datenbank und Dashboard auf einem Cloud-Server)"
            }
          ]
        }
      ]
    },
    {
      "name": "page3",
      "elements": [
        {
          "type": "radiogroup",
          "name": "battery_vendor",
          "visibleIf": "{q_devices} contains 'battery'",
          "title": {
            "default": "Von welchem Hersteller ist dein Stromspeicher?",
            "de": "Welchen Stromspeicher hast du?"
          },
          "isRequired": true,
          "choices": [
            {
              "value": "battery_senec3",
              "text": {
                "default": "SENEC",
                "de": "SENEC.Home V2.1 oder V3"
              }
            },
            {
              "value": "battery_senec4",
              "text": {
                "de": "SENEC.Home 4"
              }
            },
            {
              "value": "battery_other",
              "text": {
                "default": "anderer Hersteller",
                "de": "Speicher eines anderen Herstellers"
              }
            }
          ]
        }
      ]
    },
    {
      "name": "page4",
      "elements": [
        {
          "type": "radiogroup",
          "name": "heatpump_access",
          "visibleIf": "{q_devices} contains 'heatpump'",
          "title": "Wie wird der Stromverbrauch der Wärmepumpe gemessen?",
          "isRequired": true,
          "choices": [
            {
              "value": "heatpump_shelly",
              "text": "Über einen Shelly"
            },
            {
              "value": "heatpump_mqtt",
              "text": "Über MQTT"
            }
          ]
        }
      ]
    },
    {
      "name": "page5",
      "elements": [
        {
          "type": "radiogroup",
          "name": "wallbox_vendor",
          "visibleIf": "{q_devices} contains 'wallbox'",
          "title": "Von welchem Hersteller ist deine Wallbox?",
          "choices": [
            {
              "value": "wallbox_senec",
              "text": "SENEC (gesteuert vom Stromspeicher)"
            },
            {
              "value": "wallbox_other",
              "text": "anderer Hersteller"
            }
          ]
        }
      ]
    },
    {
      "name": "page6",
      "elements": [
        {
          "type": "boolean",
          "name": "q_updates",
          "title": {
            "de": "Möchtest du automatische Updates aktivieren?"
          },
          "defaultValue": "true"
        }
      ]
    },
    {
      "name": "page7",
      "elements": [
        {
          "type": "boolean",
          "name": "q_forecast",
          "title": {
            "de": "Ertragsvorhersage per forecast.solar?"
          },
          "defaultValue": "true"
        }
      ]
    },
    {
      "name": "page8",
      "elements": [
        {
          "type": "text",
          "name": "forecast_latitude",
          "title": {
            "de": "Latitude"
          },
          "description": {
            "de": "-90 (Süd) bis 90 (Nord)"
          },
          "inputType": "number"
        },
        {
          "type": "text",
          "name": "forecast_longitude",
          "title": {
            "de": "Longitude"
          },
          "description": {
            "de": "-180 (West) bis 180 (Ost)"
          },
          "inputType": "number"
        },
        {
          "type": "text",
          "name": "forecast_declination",
          "title": {
            "de": "Dachneigung"
          },
          "description": {
            "de": "0 (horizontal) bis 90 (Vertikal)"
          },
          "inputType": "number"
        },
        {
          "type": "text",
          "name": "forecast_azimuth",
          "title": {
            "de": "Ausrichtung"
          },
          "description": {
            "de": "-180 … 180 (-180 = Nord, -90 = Ost, 0 = Süd, 90 = West, 180 = Nord)"
          },
          "inputType": "number"
        },
        {
          "type": "text",
          "name": "forecast_kwp",
          "title": {
            "de": "Maximal-Leistung"
          },
          "description": {
            "de": "kWp"
          },
          "inputType": "number"
        },
        {
          "type": "text",
          "name": "forecast_interval",
          "title": {
            "de": "Abfrage-Intervall"
          },
          "description": {
            "de": "Angabe in Sekunden"
          },
          "inputType": "number"
        }
      ],
      "visibleIf": "{q_forecast} = true",
      "title": {
        "de": "Standort der PV-Anlage"
      },
      "description": {
        "de": "Erforderlich für die Ertragsvorhersage"
      }
    }
  ]
}
