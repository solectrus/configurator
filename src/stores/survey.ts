import { defineStore } from 'pinia'
import { Model } from 'survey-core'
import 'survey-core/i18n/german'
import { BorderlessLight } from 'survey-core/themes/borderless-light'

import { EnvGeneratorService } from '@/services/EnvGeneratorService'
import { ComposeGeneratorService } from '@/services/ComposeGeneratorService'

import type { Answers } from '@/types/answers'

interface SurveyState {
  surveyJson: any | null
  survey: Model | null
  answers: Answers | null
  finished: boolean
  composeFile: string
  envFile: string
}

export const useSurveyStore = defineStore('survey', {
  state: (): SurveyState => ({
    surveyJson: null,
    survey: null,
    answers: null,
    finished: false,
    composeFile: '',
    envFile: '',
  }),

  actions: {
    setSurvey(surveyJson: any) {
      this.surveyJson = surveyJson
      this.survey = new Model(surveyJson)
      this.survey.applyTheme(BorderlessLight)
      this.survey.locale = 'de'
      this.survey.onValueChanged.add((sender) => this.setAnswers(sender.data))
      this.survey.onComplete.add(() => (this.finished = true))
      this.setAnswers(null)

      return this.survey
    },

    setAnswers(newAnswers: Answers | null) {
      this.answers = newAnswers
      this.finished = false

      if (this.answers) {
        const compose = new ComposeGeneratorService(this.answers).build()

        this.composeFile = compose.text()
        this.envFile = new EnvGeneratorService(compose.raw(), this.answers).build()
      } else {
        this.composeFile = this.envFile = ''
      }
    },

    reset() {
      if (this.survey) {
        // this.survey.clear(true, true) does not work as expected (it does not go to the first page)
        // so we have to do it manually
        this.setSurvey(this.surveyJson)
      }
    },
  },
})
