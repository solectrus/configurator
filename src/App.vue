<script setup lang="ts">
import TheSurvey from '@/components/TheSurvey.vue'
import TheOutput from '@/components/TheOutput.vue'

import { useSurveyStore } from '@/stores/survey'

const store = useSurveyStore()
</script>

<template>
  <div class="background-image"></div>

  <div class="flex h-screen w-screen flex-col">
    <main class="flex-1 justify-center lg:flex lg:gap-10 lg:overflow-hidden">
      <TheSurvey class="no-scrollbar max-w-2xl flex-1 lg:overflow-y-scroll" />
      <TheOutput class="no-scrollbar lg:overflow-y-scroll" v-if="store.contentAvailable" />
    </main>
  </div>
</template>

<style scoped>
.background-image {
  @apply fixed -z-10 h-screen w-screen bg-cover bg-center;

  background-image: url('/images/background.jpg');
}
</style>
