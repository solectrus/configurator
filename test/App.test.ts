import { shallowMount } from '@vue/test-utils'
import { createPinia, setActivePinia } from 'pinia'
import { useSurveyStore } from '@/stores/survey'

import App from '@/App.vue'

describe('App', () => {
  setActivePinia(createPinia())
  const store = useSurveyStore()

  beforeEach(() => {
    store.initSurvey()
  })

  it('can mount with content available', () => {
    store.survey.setValue('devices', ['inverter'])
    store.survey.setValue('installation_type', 'local')
    expect(store.contentAvailable).toBe(true)

    const wrapper = shallowMount(App)
    expect(wrapper.element).toMatchSnapshot()
  })

  it('can mount without content available', () => {
    expect(store.contentAvailable).toBe(false)

    const wrapper = shallowMount(App)
    expect(wrapper.element).toMatchSnapshot()
  })
})
