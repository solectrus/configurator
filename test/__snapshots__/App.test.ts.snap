// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > can mount with content available 1`] = `
<div
  data-v-app=""
>
  
  <div
    class="background-image"
    data-v-7a7a37b1=""
  />
  <div
    class="flex h-screen w-screen flex-col"
    data-v-7a7a37b1=""
  >
    <main
      class="flex-1 justify-center lg:flex lg:gap-10 lg:overflow-hidden"
      data-v-7a7a37b1=""
    >
      <the-survey-stub
        class="no-scrollbar max-w-2xl flex-1 lg:overflow-y-scroll"
        data-v-7a7a37b1=""
      />
      <the-output-stub
        class="no-scrollbar lg:overflow-y-scroll"
        data-v-7a7a37b1=""
      />
    </main>
  </div>
  
</div>
`;

exports[`App > can mount without content available 1`] = `
<div
  data-v-app=""
>
  
  <div
    class="background-image"
    data-v-7a7a37b1=""
  />
  <div
    class="flex h-screen w-screen flex-col"
    data-v-7a7a37b1=""
  >
    <main
      class="flex-1 justify-center lg:flex lg:gap-10 lg:overflow-hidden"
      data-v-7a7a37b1=""
    >
      <the-survey-stub
        class="no-scrollbar max-w-2xl flex-1 lg:overflow-y-scroll"
        data-v-7a7a37b1=""
      />
      <!--v-if-->
    </main>
  </div>
  
</div>
`;
